solutions/usaco-718.mdx

```diff
@@ -2,29 +2,65 @@
 id: usaco-718
 source: USACO Gold 2017 February
 title: Why Did the Cow Cross The Road II
-author: Brad Ma
+author: Brad Ma, Jesse Choe
 ---
  
 [Official Analysis (C++)](http://www.usaco.org/current/data/sol_nocross_gold_feb17.html)
 
-The recursive approach where you try all configurations of crosswalks has the 
-time complexity $\mathcal{O}(2 ^ N)$, which is extremely slow. We observe that we have 
-many duplicate recursive calls where the configuration has been solved already, so by 
-using a top down dynamic progamming approach with memoization we can significantly reduce 
-the time complexity to $\mathcal{O}(N ^ 2)$. 
+For convenience, let's define $\texttt{dp}[i][j]$ represent the maximum number of disjoint friendly crosswalks up to field $i$ on the first side of the road and field $j$ on the second side of the road. Since the crosswalks must be disjoint, we can build a crosswalks as long as there isn't a crosswalk between fields $i$, $j - 1$ and $i - 1$, $j$. Thus, our transition is $dp[i][j] = max(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1] + (|id[i] - id[j]| <= 4))$, where $id[i]$ and $id[j]$ represent the breed IDs of the cows in the $i$-th and $j$-th fields, respectively.
```

Comment:
```suggestion
For convenience, let's define $\texttt{dp}[i][j]$ as the maximum number of disjoint friendly crosswalks up to field $i$ on the first side of the road and field $j$ on the second side of the road. Since the crosswalks must be disjoint, we can build a crosswalks as long as there isn't a crosswalk between fields $i$, $j - 1$ and $i - 1$, $j$. Thus, our transition is $dp[i][j] = max(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1] + (|id[i] - id[j]| <= 4))$, where $id[i]$ and $id[j]$ represent the breed IDs of the cows in the $i$-th and $j$-th fields, respectively.
```